### PR TITLE
ECMP FRR Automation test cases

### PR DESCRIPTION
**Pull request by sjwahub**
Signed-off-by: Shujun Wang <shujun.wang@intel.com>

ECMP FRR automation test cases.

1.  Testing log  [ptf_tests_run_frr_7.log](https://github.com/ipdk-io/ovs/files/9913781/ptf_tests_run_frr_7.log)

```
====================
Summary
====================
LNT_FRR_without_ECMP_ping_10min_LinkFlap : PASSED
LNT_FRR_without_ECMP_netperf : PASSED
LNT_FRR_with_ECMP_modify_TEP_IP : PASSED
LNT_FRR_with_ECMP_modify_BGP_ASN : PASSED
LNT_FRR_with_ECMP_ping_10min : PASSED
LNT_FRR_with_ECMP_netperf_LinkFlap : PASSED
LNT_FRR_with_ECMP_Hotplug_netperf : PASSED
```

2.  Building p4ovs log [BuildLog.txt](https://github.com/ipdk-io/ovs/files/9890580/BuildLog.txt)

### CODE REVIEW

**Comment by ipdksw** [comment]
Result:FAILURE JenkinsConsole:https://s3.ipdk.io/ovs/PR-84-1/index.html GitCommit:https://github.com/ipdk-io/ovs/commits/4388575878aeb182551987373f31172800febd24

File: ptf_tests/common/utils/config_file_utils.py [hunk 1]
```diff
@@ -199,16 +197,69 @@ def get_gnmi_params_simple(data):
...
                 param += f"{field}:{port[field]},"
-
+   
         param += f"port-type:{port['port-type']}"
-
         params.append(param)
 
     return params
+    
+def get_gnmi_all_port_params_simple(data):
```
Comment (dkaranjai) [inline ptf_tests/common/utils/config_file_utils.py]: The function name doesn't convey the purpose. maybe get_gnmi_phy_with_ctrl_port or something similar.

File: ptf_tests/common/utils/config_file_utils.py [hunk 2]
```diff
@@ -199,16 +197,69 @@ def get_gnmi_params_simple(data):
...
+        for field in optional:
+            if field in port.keys():
+                param += f"{field}:{port[field]},"
+                
+        param += f"port-type:{port['port-type']}" 
+        params.append(param) 
+            
+        if device_type =="link" and port.get('control-port'):
+            ctrl_port = f"device:virtual-device,name:{port['control-port']},"
+
```
Comment (dkaranjai) [inline ptf_tests/common/utils/config_file_utils.py:257]: document what you are doing here

File: ptf_tests/common/utils/gnmi_cli_utils.py [hunk 3]
```diff
@@ -97,7 +98,8 @@ def gnmi_get_params_counter(param):
          port_counter[each] = int(next(iter_rslt))
 
     if port_counter:
-           return port_counter
+        print (f"PASS: The port {param.split(',')[1].split(':')[1]} counter is built successfully")
```
Comment (dkaranjai) [inline ptf_tests/common/utils/gnmi_cli_utils.py]: can we move this print statement to the test case?

File: ptf_tests/common/utils/test_utils.py [hunk 4]
```diff
@@ -1017,7 +993,249 @@ def host_check_netperf(remote=False, hostname="",username="",password=""):
...
         return False
-    print (f"PASS: pkill -9 netserver and ready to restart netserver on {hostname}")
+    print (f"PASS: \"pkill -9 netserver\" and ready to restart netserver on {hostname}")
+    
+    connection.tear_down()
+    return True
+
+def get_vm_interface_counter(conn, interface):
+    """
+    A funtion to build a counter directory for a VM interface statics
```
Comment (dkaranjai) [inline ptf_tests/common/utils/test_utils.py]: typo "function"

Comment (dkaranjai) [reply ptf_tests/common/utils/test_utils.py]: and "statistics"

File: ptf_tests/common/utils/test_utils.py [hunk 5]
```diff
@@ -1017,7 +993,249 @@ def host_check_netperf(remote=False, hostname="",username="",password=""):
...
+    """
+    counter ={}
+    cmd= f"ip -s -s link show dev {interface}"
```
Comment (dkaranjai) [inline ptf_tests/common/utils/test_utils.py:1007]: -s twice. is this correct?

File: ptf_tests/common/utils/test_utils.py [hunk 6]
```diff
@@ -1017,7 +993,249 @@ def host_check_netperf(remote=False, hostname="",username="",password=""):
...
+            config.kill(0)
+        elif i == 1:
+            print('Connected Successfully.')
+            config.sendline("vtysh")
+            config.expect("#")
+    else:
+        config= pexpect.spawn("vtysh")
+        config.expect("#")
+    try:
+        if not asnumber:
```
Comment (dkaranjai) [inline ptf_tests/common/utils/test_utils.py]: asnumber is a mandatory argument here, by "not asnumber" do you mean asnumber=0 or asnumber="" or None?

**Review by dkaranjai — CHANGES_REQUESTED** [review]
please run the entire suite once (if not done already) since it involves lot of changes in existing files.

**Comment by ipdksw** [comment]
Result:SUCCESS JenkinsConsole:https://s3.ipdk.io/ovs/PR-84-2/index.html GitCommit:https://github.com/ipdk-io/ovs/commits/250d8927b27ac1618392b66b981ea6def0e5d9bb

**Comment by ipdksw** [comment]
Result:SUCCESS JenkinsConsole:https://s3.ipdk.io/ovs/PR-84-3/index.html GitCommit:https://github.com/ipdk-io/ovs/commits/250d8927b27ac1618392b66b981ea6def0e5d9bb

File: ptf_tests/common/utils/test_utils.py (hunk 5 above)
Comment (sjwahub) [reply ptf_tests/common/utils/test_utils.py:1007]: It works. More information provided 

[root@wp26 install]# ip -s -s link show dev eno1
2: eno1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP mode DEFAULT group default qlen 1000
    link/ether a4:bf:01:3e:1e:44 brd ff:ff:ff:ff:ff:ff
    RX: bytes  packets  errors  dropped overrun mcast
    12045185209 8544775  0       0       0       52905
    RX errors: length   crc     frame   fifo    missed
               0        0       0       0       0
    TX: bytes  packets  errors  dropped carrier collsns
    143657346  1380950  0       0       0       0
    TX errors: aborted  fifo   window heartbeat transns
               0        0       0       0       2
    altname enp61s0f0

[root@wp26 install]#
[root@wp26 install]# ip -s link show dev eno1
2: eno1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP mode DEFAULT group default qlen 1000
    link/ether a4:bf:01:3e:1e:44 brd ff:ff:ff:ff:ff:ff
    RX: bytes  packets  errors  dropped overrun mcast
    12045202149 8545027  0       0       0       52905
    TX: bytes  packets  errors  dropped carrier collsns
    143741828  1381367  0       0       0       0
    altname enp61s0f0

File: ptf_tests/common/utils/config_file_utils.py (hunk 1 above)
Comment (sjwahub) [reply ptf_tests/common/utils/config_file_utils.py]: updated according to the comment

File: ptf_tests/common/utils/config_file_utils.py (hunk 2 above)
Comment (sjwahub) [reply ptf_tests/common/utils/config_file_utils.py:257]: Done!

File: ptf_tests/common/utils/gnmi_cli_utils.py (hunk 3 above)
Comment (sjwahub) [reply ptf_tests/common/utils/gnmi_cli_utils.py]: update the code accordingly.

File: ptf_tests/common/utils/test_utils.py (hunk 4 above)
Comment (sjwahub) [reply ptf_tests/common/utils/test_utils.py]: corrected them accordingly

**Comment by ipdksw** [comment]
Result:SUCCESS JenkinsConsole:https://s3.ipdk.io/ovs/PR-84-4/index.html GitCommit:https://github.com/ipdk-io/ovs/commits/7391cd88ad27d938cda158f39a04c5682c305bc4

**Comment by ipdksw** [comment]
Result:SUCCESS JenkinsConsole:https://s3.ipdk.io/ovs/PR-84-6/index.html GitCommit:https://github.com/ipdk-io/ovs/commits/987d84cb91e29184260dfc60415fb7dc2768bb3f

**Review by monikajagia — APPROVED** [review]

File: ptf_tests/common/utils/test_utils.py (hunk 6 above)
Comment (sjwahub) [reply ptf_tests/common/utils/test_utils.py]: resolved per discussion.

**Comment by ipdksw** [comment]
Result:FAILURE JenkinsConsole:https://s3.ipdk.io/ovs/PR-84-7/index.html GitCommit:https://github.com/ipdk-io/ovs/commits/63982431103f583e901fb6673807bf103937cd4e

**Comment by sjwahub** [comment]
Final run. In order to verify this check-in won't impact on any existing case.

1. Test cases based on LNT topology along with other cases [ptf_tests_run_all_5.log](https://github.com/ipdk-io/ovs/files/9947728/ptf_tests_run_all_5.log)

- There are 4 failure cases.  None of them is related to utility changes.
- ct_multiple_connection_vhost_ports_with_netperf failure was caused by no netperf installed in VM image. Using right VM image, and rerun it as a single case. It passed as [ct_mutl_connec.txt](https://github.com/ipdk-io/ovs/files/9947752/ct_mutl_connec.txt)
-  Other failure were related to LNT netperf issue. The netperf failure happened in the middle of iteration of generating traffic. That means no logical flaw. It's known issue which I am working on to fix these intermittent problem

 
2. Test cases based on link port topology along with other cases : [ptf_link_port_related_case.log](https://github.com/ipdk-io/ovs/files/9947723/ptf_link_port_related_case.log)
3. p4c_artifact/linuxnetworking is updated.

**Comment by ipdksw** [comment]
Result:FAILURE JenkinsConsole:https://s3.ipdk.io/ovs/PR-84-8/index.html GitCommit:https://github.com/ipdk-io/ovs/commits/481d4f7b63e38232a16250253c1544bcf70a273f

**Comment by ipdksw** [comment]
Result:SUCCESS JenkinsConsole:https://s3.ipdk.io/ovs/PR-84-9/index.html GitCommit:https://github.com/ipdk-io/ovs/commits/0ebfa6a7dc7d0920b52de3e2186c740788f8188e

**Review by dkaranjai — APPROVED** [review]
LGTM

**Pull request merged**